### PR TITLE
Mutual TLS Auth server and clients

### PR DESCRIPTION
**Pull request by klizhentas**
## Description

This PR replaces SSH based Auth server authentication with TLS based mutual authentication. This PR is a first in a series of follow ups but is self-sufficient.

## Included in the PR:

* 2.5.0 Server will serve SSH and TLS on the same auth server socket serving legacy SSH client requests and TLS requests
* Trusted clusters will upgrade and re-exchange certificates after upgrade
* Nodes will re-register and receive new identities after auth server upgrade
* Auth server switched to use mutual TLS client authentication
* Added support for proxy line protocol. The protocol support is turned on by default, but can be turned off later.
* No reconfiguration required, upgrade is fully automatic.

## Not included in the PR

* Some edge cases with additional tests
* User/Admin upgrade and install documentation

### CODE REVIEW

File: lib/auth/register.go [hunk 1]
```diff
@@ -45,30 +49,56 @@ func Register(dataDir, token string, id IdentityID, servers []utils.NetAddr) err
...
+		}
+		message := fmt.Sprintf(`Your configuration is insecure! Registering without TLS certificate authority, to fix this warning add ca.cert to %v, you can get ca.cert using 'tctl auth export --type=tls > ca.cert'`, dataDir)
+		log.Warning(message)
+		tlsConfig.InsecureSkipVerify = true
+	} else {
+		cert, err := tlsca.ParseCertificatePEM(certBytes)
+		if err != nil {
+			return trace.Wrap(err, "failed to parse certificate at %v", certPath)
+		}
+		log.Infof("Securely joining remote cluster %v.", cert.Subject.CommonName)
```
Comment (klizhentas) [inline lib/auth/register.go]: remove Securely

File: lib/client/client.go [hunk 2]
```diff
@@ -262,6 +275,56 @@ func (proxy *ProxyClient) isRecordingProxy() (bool, error) {
 	}
 }
 
+// dialAuthServer returns auth server connection forwarded via proxy
+func (proxy *ProxyClient) dialAuthServer(ctx context.Context, clusterName string) (net.Conn, error) {
+	log.Infof("[CLIENT] client=%v connecting to auth server on cluster ", proxy.clientAddr, clusterName)
```
Comment (klizhentas) [inline lib/client/client.go]: remove square brackets

File: lib/auth/init.go [hunk 3]
```diff
@@ -440,8 +436,7 @@ func migrateRoles(asrv *AuthServer) error {
 		if err != nil {
 			return trace.Wrap(err)
 		}
-
-		log.Infof("[MIGRATION] Updated Role: %v to include port_forwarding and forward_agent option", role.GetName())
+		log.Infof("Migrating role: %v to include port_forwarding and forward_agent option", role.GetName())
```
Comment (r0mant) [inline lib/auth/init.go]: Dot at the end.

File: lib/auth/init.go [hunk 4]
```diff
@@ -470,25 +465,34 @@ func isFirstStart(authServer *AuthServer, cfg InitConfig) (bool, error) {
...
 	}
 
-	certExists, err := pathExists(cp)
+	tlsCertExists, err := pathExists(path.tlsCert)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	if !keyExists || !certExists {
+	if !keyExists || !sshCertExists || !tlsCertExists {
```
Comment (r0mant) [inline lib/auth/init.go]: Won't this cause SSH cert and key to be overwritten for existing clusters that do not have TLS cert yet?

File: lib/auth/init.go [hunk 5]
```diff
@@ -601,67 +720,99 @@ func ReadIdentityFromKeyPair(keyBytes, certBytes []byte) (*Identity, error) {
 			foundRoles, roles.String())
 	}
 	role := roles[0]
-	authorityDomain := cert.Permissions.Extensions[utils.CertExtensionAuthority]
-	if authorityDomain == "" {
+	clusterName := cert.Permissions.Extensions[utils.CertExtensionAuthority]
+	if clusterName == "" {
 		return nil, trace.BadParameter("misssing cert extension %v", utils.CertExtensionAuthority)
```
Comment (r0mant) [inline lib/auth/init.go]: sss -> ss

File: lib/auth/methods.go [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+		if authPreference.GetSecondFactor() != teleport.OFF {
+			return trace.AccessDenied("missing second factor")
+		}
+		err := s.WithUserLock(req.Username, func() error {
+			return s.CheckPasswordWOToken(req.Username, req.Pass.Password)
+		})
+		if err != nil {
+			// provide obscure message on purpose, while logging the real
+			// error server side
+			log.Debugf("failed to authenticate: %v", err)
```
Comment (r0mant) [inline lib/auth/methods.go]: "Failed", dot at the end, trace.DebugReport.

File: lib/auth/methods.go [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+	case req.U2F != nil:
+		// authenticate using U2F - code checks challenge response
+		// signed by U2F device of the user
+		err := s.WithUserLock(req.Username, func() error {
+			return s.CheckU2FSignResponse(req.Username, &req.U2F.SignResponse)
+		})
+		if err != nil {
+			// provide obscure message on purpose, while logging the real
+			// error server side
+			log.Debugf("failed to authenticate: %v", err)
```
Comment (r0mant) [inline lib/auth/methods.go]: "Failed", dot at the end, trace.DebugReport.

File: lib/auth/methods.go [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+		}
+		return nil
+	case req.OTP != nil:
+		err := s.WithUserLock(req.Username, func() error {
+			return s.CheckPassword(req.Username, req.OTP.Password, req.OTP.Token)
+		})
+		if err != nil {
+			// provide obscure message on purpose, while logging the real
+			// error server side
+			log.Debugf("failed to authenticate: %v", err)
```
Comment (r0mant) [inline lib/auth/methods.go]: "Failed", dot at the end, trace.DebugReport. You could probably move the error handling outside the switch as all methods handle it same way.

File: lib/auth/methods.go [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+			ClusterName:      ca.GetClusterName(),
+			HostCertificates: ca.GetCheckingKeys(),
+			TLSCertificates:  services.TLSCerts(ca),
+		}
+	}
+	return out
+}
+
+// AuthenticateSSHUser authenticates web user, creates and  returns web session
+// in case if authentication is successfull
```
Comment (r0mant) [inline lib/auth/methods.go]: successful<strike>l</strike>

File: lib/client/keystore.go [hunk 10]
```diff
@@ -200,22 +220,85 @@ func (fs *FSLocalKeyStore) GetKey(host, username string) (*Key, error) {
...
+	}
+	pool := x509.NewCertPool()
+	for len(bytes) > 0 {
+		var block *pem.Block
+		block, bytes = pem.Decode(bytes)
+		if block == nil {
+			break
+		}
+		if block.Type != "CERTIFICATE" || len(block.Headers) != 0 {
+			log.Debugf("skipping PEM block type=%v headers=%v", block.Type, block.Headers)
```
Comment (r0mant) [inline lib/client/keystore.go]: "Skipping", dot at the end.

File: lib/client/keystore.go [hunk 11]
```diff
@@ -200,22 +220,85 @@ func (fs *FSLocalKeyStore) GetKey(host, username string) (*Key, error) {
...
+			continue
+		}
+
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return nil, trace.BadParameter("failed to parse certificate: %v", err)
+		}
+		log.Debugf("Addding trusted cluster certificate authority %q to trusted pool.", cert.Issuer)
```
Comment (r0mant) [inline lib/client/keystore.go]: ddd -> dd

File: lib/reversetunnel/remotesite.go [hunk 12]
```diff
@@ -269,6 +317,73 @@ func (s *remoteSite) periodicSendDiscoveryRequests() {
...
+		DomainName: s.domainName,
+	}, false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	_, err = s.localClient.ExchangeCerts(auth.ExchangeCertsRequest{
+		PublicKey: remoteCA.GetCheckingKeys()[0],
+		TLSCert:   re.TLSCert,
+	})
+	return err
```
Comment (r0mant) [inline lib/reversetunnel/remotesite.go]: trace.Wrap

File: lib/service/service.go [hunk 13]
```diff
@@ -179,24 +200,48 @@ func (process *TeleportProcess) connectToAuthService(role teleport.Role) (*Conne
...
+	tlsConfig, err := identity.TLSConfig()
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return nil, trace.Wrap(err)
+		}
+		// connect using legacy SSH and get new set of TLS credentials
+		storage := utils.NewFileAddrStorage(
+			filepath.Join(process.Config.DataDir, "authservers.json"))
+
+		authUser := identity.Cert.ValidPrincipals[0]
```
Comment (r0mant) [inline lib/service/service.go]: Is the `ValidPrincipals` length checked anywhere?

File: lib/services/local/presence.go [hunk 14]
```diff
@@ -473,10 +473,15 @@ func (s *PresenceService) DeleteTunnelConnections(clusterName string) error {
 
 // DeleteAllTunnelConnections deletes all tunnel connections
 func (s *PresenceService) DeleteAllTunnelConnections() error {
+	conns, _ := s.GetAllTunnelConnections()
+	log.Debugf("Presence delete all tunnel connections: %v", conns)
 	err := s.DeleteBucket([]string{}, tunnelConnectionsPrefix)
+	log.Debugf("Presence delete all tunnel connections: %v", err)
```
Comment (r0mant) [inline lib/services/local/presence.go]: Log entries in this function look like debug leftovers.

File: lib/services/local/presence.go [hunk 15]
```diff
@@ -473,10 +473,15 @@ func (s *PresenceService) DeleteTunnelConnections(clusterName string) error {
 
 // DeleteAllTunnelConnections deletes all tunnel connections
 func (s *PresenceService) DeleteAllTunnelConnections() error {
+	conns, _ := s.GetAllTunnelConnections()
```
Comment (r0mant) [inline lib/services/local/presence.go]: Error is ignored.

File: lib/services/local/presence.go [hunk 16]
```diff
@@ -473,10 +473,15 @@ func (s *PresenceService) DeleteTunnelConnections(clusterName string) error {
...
+	log.Debugf("Presence delete all tunnel connections: %v", conns)
 	err := s.DeleteBucket([]string{}, tunnelConnectionsPrefix)
+	log.Debugf("Presence delete all tunnel connections: %v", err)
 	if trace.IsNotFound(err) {
 		return nil
 	}
+	conns, _ = s.GetAllTunnelConnections()
```
Comment (r0mant) [inline lib/services/local/presence.go]: Error is ignored. This actually, too, looks like a debug leftover. Getting tunnels in "delete" method?

File: lib/state/cachingaccesspoint.go [hunk 17]
```diff
@@ -398,6 +398,30 @@ func (cs *CachingAuthClient) GetProxies() (proxies []services.Server, err error)
...
+	if err := cs.trust.DeleteCertAuthority(id); err != nil {
+		if !trace.IsNotFound(err) {
+			return nil, trace.Wrap(err)
+		}
+	}
+	cs.setTTL(ca)
+	if err := cs.trust.UpsertCertAuthority(ca); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return ca, err
```
Comment (r0mant) [inline lib/state/cachingaccesspoint.go]: `return ca, nil`?

**Review by r0mant — APPROVED** [review]
Mostly typos and inconsistent log entries.

File: lib/auth/init.go (hunk 4 above)
Comment (klizhentas) [reply lib/auth/init.go]: yes, but that's okay

File: lib/service/service.go (hunk 13 above)
Comment (klizhentas) [reply lib/service/service.go]: yes auth/init.go:696

**Review by russjones — APPROVED** [review]

**Comment by klizhentas** [comment]
closing it in favor of another branch with merged conflicts

**Pull request closed without merging**